include/nanobind/nb_class.h

```diff
@@ -376,6 +376,29 @@ template <typename Arg> struct init_implicit {
     }
 };
 
+template <typename Func, typename Sig = detail::function_signature_t<Func>>
+struct new_;
+
+template <typename Func, typename Return, typename... Args>
+struct new_<Func, Return(Args...)> {
+    std::remove_reference_t<Func> func;
+
+    new_(Func &&f) : func((detail::forward_t<Func>) f) {}
+
+    template <typename Class, typename... Extra>
+    NB_INLINE void execute(Class &cl, const Extra&... extra) {
+        using Type = typename Class::Type;
+        cl.def_static(
+            "__new__",
+            [func = (detail::forward_t<Func>) func](type_object_t<Type>, Args... args) {
```

Comment:
Do you need to take `type_object_t` as first argument? You aren't really using this argument at all, and this will add an extra check done by nanobind. The ``__new__`` signature is not added to the stubs.